renderdoc/driver/ihv/amd/amd_counters.cpp

```diff
@@ -101,34 +233,61 @@ AMDCounters::~AMDCounters()
 {
   if(m_pGPUPerfAPI)
   {
-    GPA_Status status = GPA_STATUS_OK;
+    auto status = m_pGPUPerfAPI->GPA_CloseContext();
+    if (AMD_FAILED(status))
+    {
+      GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Close context failed", status).c_str());
+    }
 
-    status = m_pGPUPerfAPI->GPA_CloseContext();
     status = m_pGPUPerfAPI->GPA_Destroy();
+    if (AMD_FAILED(status))
+    {
+      GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Destroy failed", status).c_str());
+    }
 
     delete m_pGPUPerfAPI;
   }
 }
 
-vector<AMDCounters::InternalCounterDescription> AMDCounters::EnumerateCounters()
+std::map<uint32_t, CounterDescription> AMDCounters::EnumerateCounters()
 {
-  gpa_uint32 num;
+  std::map<uint32_t, CounterDescription> counters;
 
-  m_pGPUPerfAPI->GPA_GetNumCounters(&num);
+  gpa_uint32 num;
 
-  vector<InternalCounterDescription> counters;
+  auto status = m_pGPUPerfAPI->GPA_GetNumCounters(&num);
+  if (AMD_FAILED(status))
+  {
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Get number of counters", status).c_str());
+    return counters;
+  }
 
   for(uint32_t i = 0; i < num; ++i)
   {
-    InternalCounterDescription internalDesc;
-    internalDesc.desc = InternalGetCounterDescription(i);
+    GPA_Usage_Type usageType;
+
+    status = m_pGPUPerfAPI->GPA_GetCounterUsageType(i, &usageType);
+    if (AMD_FAILED(status))
+    {
+      GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Get counter usage type.", status).c_str());
+      return counters;
+    }
 
-    internalDesc.internalIndex = i;
-    internalDesc.desc.counter = MakeAMDCounter(i);
-    counters.push_back(internalDesc);
+    // Ignore percentage counters due to aggregate roll-up support
+    if (usageType == GPA_USAGE_TYPE_PERCENTAGE)
+    {
+      continue;
+    }
+
+    CounterDescription desc = InternalGetCounterDescription(i);
+
+    desc.counter = MakeAMDCounter(i);
+    counters[i] = desc;
+
+    m_PublicToInternalCounter[desc.counter] = i;
   }
 
-  return counters;
+  return std::move(counters);
```

Comment:
I'm not sure why the use of `std::move` here, since the function is still returning a temporary rather than a move reference. If you're worried about the copy cost I'd say instead define the function as `void AMDCounters::EnumerateCounters()` and write directly to `m_Counters` instead of a temporary, since with the addition of `m_PublicToInternalCounter` it's not re-usable somewhere else anyway.

Reply:
Previously fixed